WIP CRL benchmarks

cpu opened the pull request:
## Description

This branch has some initial work-in-progress for benchmarking CRL support. Sharing in a rough early form so that the results can be used to evaluate the direction we take with other work in progress code.

I'll separate out the DER support fix into another branch but needed it here to be able to perform meaningful benchmarking.

I don't think this is quite ready to go yet: GitHub is (rightfully) complaining about the CRL that's larger than 50mb. We'll want to generate that test data locally (either with Python, or `rcgen` once it supports CRL generation) as opposed to bundling it into the repo.

### der/crl: fix support for large CRL DER.
Previously we updated the `der` module to have methods that allow specifying an explicit size limit so that the `crl` module could opt-in to reading DER values larger than the implicit maximum supported by *ring*.

Unfortunately this update wasn't quite complete: we need a `der::nested_limited` version of `der::nested` to read the nested sequence with the large signed data we pass to `signed_data::parse_signed_data` or we'll run into the implicit limit before we get to the parsing that supports a larger data size.

Similarly when reading the `SEQUENCE` that holds all of the revoked certificates we need to use `der::expect_tag_and_get_value_limited` with a large limit instead of using `der::expect_tag_and_get_value` with the implicit limit. This sequence is expected to be quite large when there are a large number of revoked certs.

### benches: initial CRL parsing/searching benchmarks.
This commit adds two kinds of CRL benchmark tests:

1. Parsing benchmarks, testing how long it takes to use `CertRevocationList::from_der`.
2. Searching benchmarks, testing how long it takes to search an already parsed CRL for a serial number that doesn't appear in the CRL (forcing a full scan) using `CertRevocationList.find_serial`.

In both cases we add benchmarks that use three sizes of CRLs:

1. a tiny 5mb CRL.
2. a medium size 22mb CRL.
3. a chonky 56mb CRL.

It's likely better to generate these CRLs locally as required as opposed to checking in a large ~56mb binary file to the repo. I will experiment for a solution to that in a subsequent commit.

Initial results from my laptop:
```
running 6 tests
test bench_parsecrl_22mb  ... bench:  14,330,173 ns/iter (+/- 2,884,813)
test bench_parsecrl_56mb  ... bench:  35,010,006 ns/iter (+/- 3,282,729)
test bench_parsecrl_5mb   ... bench:   4,078,951 ns/iter (+/- 927,771)
test bench_searchcrl_22mb ... bench:  14,018,128 ns/iter (+/- 931,067)
test bench_searchcrl_56mb ... bench:  34,883,072 ns/iter (+/- 6,016,241)
test bench_searchcrl_5mb  ... bench:   3,924,192 ns/iter (+/- 157,235)
```

Laptop specs:
* rustc 1.69.0
* NixOS 22.11
* 12th Gen Intel(R) Core(TM) i7-1280P
* 32GB DDR4-3200
* WD_BLACK SN850 NVMe

djc commented:
> Unfortunately this update wasn't quite complete: we need a `der::nested_limited` version of `der::nested` to read the nested sequence with the large signed data we pass to `signed_data::parse_signed_data` or we'll run into the implicit limit before we get to the parsing that supports a larger data size.

Would it be better to have a single `nested` with an `Option` argument for `limit`, so that every caller is forced to think whether it needs a limit?

cpu commented:
> Should be fixed after rebasing on top of https://github.com/rustls/webpki/pull/95.

Thanks! That's a nice fix.

> Would it be better to have a single nested with an Option argument for limit, so that every caller is forced to think whether it needs a limit?

I think @ctz's thought process on the other PRs was that the size limit enforcement isn't especially necessary since the caller always has to have loaded the data they feed to webpki themselves. The only codepath where the default limit is known to be inappropriate is for the new CRL code and I don't expect we'll end up wanting to fiddle with the limit elsewhere. With those two aspects in mind it felt better to avoid an `Option` argument for the default case. WDYT?

> If the goal is to prevent bitrotting, we can make sure to check (--all-targets would do it) without the need for running them to avoid the CI runtime penalty.

Sounds good, I will rework accordingly.

djc commented:
> I think @ctz's thought process on the other PRs was that the size limit enforcement isn't especially necessary since the caller always has to have loaded the data they feed to webpki themselves. The only codepath where the default limit is known to be inappropriate is for the new CRL code and I don't expect we'll end up wanting to fiddle with the limit elsewhere. With those two aspects in mind it felt better to avoid an `Option` argument for the default case. WDYT?

Makes sense.

cpu commented:
> If the goal is to prevent bitrotting, we can make sure to check (--all-targets would do it) without the need for running them to avoid the CI runtime penalty.

I verified locally that a breaking change in `benches/benchmark.rs` was caught by the existing `cargo fmt --all -- --check`: https://github.com/rustls/webpki/blob/ea200de9d2881633459d07e6aaa4c87835a75e05/.github/workflows/ci.yml#L20

I'll drop the new CI task that actually runs the benchmark.

cpu commented:
Force pushed with some updates:

* Rebased on `main` to pick up the MSRV check fix Djc applied (thanks!)
* Dropped the addition that ran `cargo bench` in CI after verifying our existing CI tasks will prevent bitrot in the benchmark code.
* Added a new commit (xxx) before the benchmarking that drops the up-front revoked certs parsing at CRL-parse time. We can discuss whether we want to keep this behaviour but it was muddying the benchmark results.

PR description updated accordingly.

The new results clearly show that CRL parsing (without revoked cert iteration) is _very_ fast.

cpu commented:
> Added a new commit (https://github.com/rustls/webpki/commit/b1c4ecf0380c7b7060fd6693e139d03446dac227) before the benchmarking that drops the up-front revoked certs parsing at CRL-parse time.

```
error[E0412]: cannot find type `Vec` in this scope
   --> src/crl.rs:229:31
    |
229 |             .collect::<Result<Vec<_>, _>>()?
    |                               ^^^ not found in this scope
```

Right... Can't alloc. Will rework.

cpu commented:
Closing this since we're going to rework the webpki representation. I'll revisit when there is new code to benchmark.
